File: front-end/src/utils/interface/response.ts
export interface Holiday {
  id: string;              // int64 ใน Go มักจะใช้ number ใน TS
  updatedAt: string; // *time.Time เป็น pointer และเป็น format ISO string
  date: string;            // DateRes (gorm:"type:date") มักส่งมาเป็น "YYYY-MM-DD"
  name: string;
  isDayOff: boolean; // *bool ใน Go สามารถเป็น true, false หรือ null ได้
  source: string;
}

export interface RoomResponse {
  id: string
  name: string
  roomNumber: number
  location: string
  capacity: number
  status: string
}

export interface UserResponse {
  id: string
  email: string
  fullName: string
  avatarUrl?: string
  role?: string
  status: string
}

export interface BookingEventResponse {
  id: string
  startTime: string
  endTime: string
  title: string
  passcode: string
  status: string
  User: UserResponse
  Room: RoomResponse
}

export interface AttendanceHealthResponse {
  completed: number;
  cancelled: number;
  noShow: number;
  completionRate: number;
  canCelledRate: number
	noShowRate: number
}

export interface PopularRoomResponseonse {
  id: number;
  roomNumber: string;
  name: string;
  percentage: number;
}

export interface DashboardAnalyticsResponse {
  attendanceHealth: AttendanceHealthResponse;
  popularRooms: PopularRoomResponseonse[];
}

export interface ConfigResponse {
  startTime: string
  endTime: string
  maxAdvanceDays: number
  maxBookingMins: number
  noShowThresholdMins: number
  earlyCheckInMinutes: number
}

export interface PaginationMeta {
  totalItems: number;       // int64 ใน Go แปลงเป็น number
  itemsPerPage: number;
  totalPages: number;
  currentPage: number;
  indexOfFirstItem: number; // ⭐️ ดึงมาจาก Struct ใหม่
  indexOfLastItem: number;  // ⭐️ ดึงมาจาก Struct ใหม่
}

// ==========================================
// 3. User & Statistics (ข้อมูลผู้ใช้และสถิติ)
// ==========================================

export interface UserStatsRes {
  upcoming: number;  // int64 แปลงเป็น number
  completed: number; 
  cancelled: number; 
  noShow: number;    
}

export interface UserOverviewResponse {
  user: UserResponse;
  statistics: UserStatsRes;
}

export interface PaginatedUserResponse {
  data: UserResponse[]; // สันนิษฐานว่าใช้โครงสร้างเดียวกับ UserInfoRes
  meta: PaginationMeta;
}

// ==========================================
// 4. Booking History & Rooms (ประวัติการจองและห้อง)
// ==========================================
export interface UserRoomRes {
  id: string;        // uuid.UUID แปลงเป็น string
  name: string;
  roomNumber: number; // uint แปลงเป็น number
  location: string;
}

export interface UserBookingHistoryRes {
  id: string;                 // uuid.UUID แปลงเป็น string
  title: string;
  startTime: string;   // *time.Time เป็น pointer เผื่อค่า null
  endTime: string;     // *time.Time เป็น pointer เผื่อค่า null
  status: string;      // *string เป็น pointer เผื่อค่า null
  checkedInAt: string; // *time.Time เป็น pointer เผื่อค่า null
  Room: UserRoomRes;          // ตัวพิมพ์ใหญ่ตาม json:"Room"
}

export interface PaginatedBookingResponse {
  data: UserBookingHistoryRes[];
  meta: PaginationMeta;
}

File: back-end/internal/domain/response.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

// Response
type WebSocketPayload struct {
	Type string 						`json:"type"`
	Data struct {
		RoomID 			uuid.UUID `json:"room_id"`
		RoomNumber 	uint 			`json:"room_number"`
		UserID     	string 		`json:"user_id"`
		Status			bool			`json:"status"`
	} 											`json:"data"`
}

type WSMessage struct {
	Type  string      `json:"type"`  // "auth" หรือ "chat" หรือ "command"
	Token string      `json:"token,omitempty"` // ใช้ตอน auth
	Data  interface{} `json:"data,omitempty"`  // ข้อมูลจริงๆ
}

// User booking
type BookingPaginationQuery struct {
	Page   int    `query:"page"`
	Limit  int    `query:"limit"`
	Status string `query:"status"` // เช่น "CONFIRMED", "COMPLETED", "ALL" (หรือปล่อยว่าง)
	Year   int    `query:"year"`   // เช่น 2026
	Month  int    `query:"month"`  // เช่น 4 (เมษายน)
}

// 📤 2. Struct ส่งออกเป็น JSON
type PaginatedBookingResponse struct {
	Data []UserBookingHistoryRes `json:"data"`
	Meta PaginationMeta          `json:"meta"`
}

// 🌟 4. Struct สำหรับตารางประวัติการจอง
type UserBookingHistoryRes struct {
	ID          uuid.UUID   `json:"id"`
	Title       string      `json:"title"`
	StartTime   *time.Time  `json:"startTime"`
	EndTime     *time.Time  `json:"endTime"`
	Status      *string     `json:"status"`
	CheckedInAt *time.Time  `json:"checkedInAt"`
	Room        UserRoomRes `json:"Room"` // ใช้ตัวพิมพ์ใหญ่ตาม JSON แบบเดิม
}

// 🌟 5. Struct สำหรับข้อมูลห้องย่อยๆ ที่อยู่ในการจอง
type UserRoomRes struct {
	ID         uuid.UUID `json:"id"`
	Name       string    `json:"name"`
	RoomNumber uint      `json:"roomNumber"`
	Location   string    `json:"location"`
}

// User & Statistic
type UserOverviewResponse struct {
	User       UserInfoRes  `json:"user"`
	Statistics UserStatsRes `json:"statistics"`
}

type UserInfoRes struct {
	ID        uuid.UUID `json:"id"`
	FullName  string    `json:"fullName"`
	Email     string    `json:"email"`
	AvatarUrl string    `json:"avatarUrl"`
	Role      string    `json:"role"`
	Status    string    `json:"status"`
}

type UserStatsRes struct {
	Upcoming  int64 `json:"upcoming"`
	Completed int64 `json:"completed"`
	Cancelled int64 `json:"cancelled"`
	NoShow    int64 `json:"noShow"`
}

// User management 
// เอาไว้รับ Query จาก URL
type UserPaginationQuery struct {
	Page   int    `query:"page"`
	Limit  int    `query:"limit"`
	Search string `query:"search"`
}

// เอาไว้ส่งกลับไปให้ Frontend
type PaginatedUserResponse struct {
	Data []User `json:"data"`
	Meta PaginationMeta `json:"meta"`
}

type PaginationMeta struct {
	TotalItems       int64 `json:"totalItems"`
	ItemsPerPage     int   `json:"itemsPerPage"`
	TotalPages       int   `json:"totalPages"`
	CurrentPage      int   `json:"currentPage"`
	IndexOfFirstItem int   `json:"indexOfFirstItem"` // ⭐️ เพิ่มเข้ามา
	IndexOfLastItem  int   `json:"indexOfLastItem"`  // ⭐️ เพิ่มเข้ามา
}